spec/light-client/Blockchain.tla

```diff
@@ -0,0 +1,212 @@
+----------------------------- MODULE Blockchain -----------------------------
+(* This is a high-level specification of Tendermint blockchain
+   that is designed specifically for:
+   
+   (1) Lite client, and
+   (2) Fork accountability.
+ *)
+EXTENDS Integers, Sequences
+
+Min(a, b) == IF a < b THEN a ELSE b
+
+CONSTANT
+  AllNodes,
+    (* a set of all nodes that can act as validators (correct and faulty) *)
+  ULTIMATE_HEIGHT,
+    (* a maximal height that can be ever reached (modelling artifact) *)
+  MAX_POWER
+    (* a maximal voting power of a single node *)
+
+Heights == 0..ULTIMATE_HEIGHT   (* possible heights *)
+
+Powers == 1..MAX_POWER          (* possible voting powers *)
+
+(* A commit is just a set of nodes who have committed the block *)
+Commits == SUBSET AllNodes
+
+(* The set of all block headers that can be on the blockchain.
+   This is a simplified version of the Block data structure in the actual implementation. *)
+BlockHeaders == [
+  height: Heights,
+    \* the block height
+  lastCommit: Commits,
+    \* the nodes who have voted on the previous block, the set itself instead of a hash
+  (* in the implementation, only the hashes of V and NextV are stored in a block,
+     as V and NextV are stored in the application state *) 
+  VP: UNION {Nodes \in SUBSET AllNodes \ {{}}: [Nodes -> Powers]},
+    \* the validators of this block together with their voting powers,
+    \* i.e., a multi-set. We store the validators instead of the hash.
+  NextVP: UNION {Nodes \in SUBSET AllNodes \ {{}}: [Nodes -> Powers]}
+    \* the validators of the next block together with their voting powers,
+    \* i.e., a multi-set. We store the next validators instead of the hash.
+]
+
+(* A signed header is just a header together with a set of commits *)
```

Comment:
In this case it is a set of commits from the next block header. Nothing to change on modelling side, maybe just to make comment more clear.

Reply:
As far as I understand it is not necessarily from the next block header. If the next block is not decided yet, a full node can return its local view of the commits, right?

Reply:
It is more like a type of signed headers, not always a block from the blockchain